packages/compiler/src/render3/view/template.ts

```diff
@@ -358,8 +358,14 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
         (scope: BindingScope, relativeLevel: number) => {
           let rhs: o.Expression;
           if (scope.bindingLevel === retrievalLevel) {
-            // e.g. ctx
-            rhs = o.variable(CONTEXT_NAME);
+            if (scope.isListenerScope()) {
+              // Either ctx or r0[CONTEXT] where r0 = getCurrentView();
+              rhs = scope.getRestoreViewVariable()?.key(o.literal(core.CONTEXT)) ||
+                  o.variable(CONTEXT_NAME);
```

Comment:
I believe we can use `??` here:
```suggestion
              rhs = scope.getRestoreViewVariable()?.key(o.literal(core.CONTEXT)) ??
                  o.variable(CONTEXT_NAME);
```